File: Software Development Module/Revision/Exam2018/SentanceEncoder.java
/*
SentanceEncoder.java
Billy Cussen
02/09/2020
*/

public class SentanceEncoder{

    //Data Members
    private StringBuffer outputSentance;

    //Constructor
    public SentanceEncoder(){
        this.outputSentance = new StringBuffer();
    }

    //Compute & Getter
    public void getEncodedSentance(String inputSentance){
        for(int i = inputSentance.length()-1; i >= 0; i--){
            if(((i+1) % 4 == 0)  && !(inputSentance.charAt(i)==' '|| inputSentance.charAt(i)=='.')){
                this.outputSentance.append("!");
            } else if(inputSentance.charAt(i)==' '){
                this.outputSentance.append("+");
            } else{
                this.outputSentance.append(inputSentance.charAt(i));
            }
        }

        System.out.println("Encoded Sentance is: "+this.outputSentance.toString());
        getCharacterCount(this.outputSentance.toString());
    }

    private void getCharacterCount(String inputSentance){
        int exclamationCount = 0, plusCount = 0;
        for(int i = 0; i < inputSentance.length(); i++){
            if(inputSentance.charAt(i)=='!'){
                exclamationCount++;
            } else if(inputSentance.charAt(i)=='+'){
                plusCount++;
            }
        }
        System.out.println("The total number of Exclamation Marks are: "+exclamationCount);
        System.out.println("The total number of Plus Characters are: "+plusCount);
    }

}

File: Python/Software Development Module - Python/Revision/Exam2018/SentanceEncoder/SentanceEncoder.py
"""
SentanceEncoder.py
Billy Cussen
02/11/2020
"""

class mySentanceEncoder:

    def __init__(self, sentance):
        self.sentance = sentance

    def getEncodedSentance(self):
        index = 0
        new_sentance = ""
        for c in reversed(self.sentance):
            if c == ' ':
                new_sentance = new_sentance + "+"
            elif (index+1) % 4 == 0 and not (c == ' ' or c == '.'):
                new_sentance = new_sentance + "!"
            else:
                new_sentance = new_sentance + c
            index = index + 1
            
        print("Encoded Sentance is: "+new_sentance)
        self.getCharacterCount(new_sentance)

    def getCharacterCount(self, new_sentance):
        exclamation_count = 0
        plus_count = 0
        for c in new_sentance:
            if c == '!':
                exclamation_count = exclamation_count + 1
            elif c == '+':
                plus_count = plus_count + 1

        print("Total number of Exclamation Marks: "+str(exclamation_count))
        print("Total number of Plus Signs: "+str(plus_count))